### ISSUE TITLE
Efficient allreduce is not supported Issue

### ISSUE THREAD

**Issue by caffeine-lab** [tensorflow/tensorflow#39545]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2 (stable)
- Python version: 3.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.1 / 7.6.4.38
- GPU model and memory: Nvidia Tesla V100 /32GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`
v2.2.0-rc4-8-g2b96f3662b 2.2.0

**Describe the current behavior**
I have written a custom loop to train a model using tf.strategy.MirroredStrategy().
Currently, I am training on a machine, equipped with 2 Nvidia V100 Tesla cards. The training is started and continuing as expected. But t I am getting a Warning from both of the GPUS, I feel this can harm the performance of the GPUs

2020-05-14 15:04:14.880452: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1247] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 30261 MB memory) -> physical GPU (device: 0, name: Tesla V100-PCIE-32GB, pci bus id: 0000:40:00.0, compute capability: 7.0)
2020-05-14 15:04:14.883196: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1247] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 30261 MB memory) -> physical GPU (device: 1, name: Tesla V100-PCIE-32GB, pci bus id: 0000:62:00.0, compute capability: 7.0)
WARNING:tensorflow:Efficient allreduce is not supported for 1 IndexedSlices
WARNING:tensorflow:Efficient allreduce is not supported for 1 IndexedSlices
2020-05-14 15:05:55.050229: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10

**Comment by amahendrakar** [comment]
@Nixon59-lab,
In order to expedite the trouble-shooting process, could you please provide the complete code and the dataset you are using to reproduce the issue reported here. Thanks!

**Comment by Nixon59-lab** [comment]
Original Script and dataset can be found here:
https://github.com/tensorflow/docs/blob/master/site/en/tutorials/text/image_captioning.ipynb

Modified custom loop:
```
@tf.function
def train_step(dist_inputs):
	def step_fn(inputs):
		img_tensor, target = inputs
		print(target.shape[0])
		loss = 0

		# initializing the hidden state for each batch because the captions are not related from image to image
		hidden = decoder.reset_state(batch_size=target.shape[0])

		dec_input = tf.expand_dims([tokenizer.word_index['<start>']] * target.shape[0], 1)

		with tf.GradientTape() as tape:
			features = encoder(img_tensor)

			for i in range(1, target.shape[1]):
				# passing the features through the decoder
				predictions, hidden, _ = decoder(dec_input, features, hidden)

				loss += loss_function(target[:, i], predictions)
				# using teacher forcing
				dec_input = tf.expand_dims(target[:, i], 1)

		total_loss = (loss / int(target.shape[1]))
		trainable_variables = encoder.trainable_variables + decoder.trainable_variables
		gradients = tape.gradient(loss, trainable_variables)
		optimizer.apply_gradients(zip(gradients, trainable_variables))
		return loss, total_loss

	per_replica_losses, l_loss = strategy.run(step_fn, args=(dist_inputs,))
	return strategy.reduce(tf.distribute.ReduceOp.MEAN, per_replica_losses,axis=None),strategy.reduce(tf.distribute.ReduceOp.MEAN, l_loss,axis=None)


for epoch in range(start_epoch, EPOCHS):
	start = time.time()
	total_loss = 0.0
	batch = 0
	for x in train_dist_dataset:
		batch_loss, t_loss = train_step(x)
		total_loss += t_loss
		batch +=1

		if batch % 100 == 0:
			print ('Epoch {} Batch {} Loss {:.4f}'.format(epoch + 1, batch, batch_loss.numpy() / BATCH_SIZE), flush=True)

	train_loss = total_loss / batch

**Comment by byronyi** [comment]
IndexedSlice are sparse data structures, so TF is using allgather for AllReduce. It probably doesn’t matter if you are only using 2 GPUs, but you could consider convert sparse tensor to a dense one for better AllReduce support.

cc @dubey

**Comment by dubey** [comment]
Actually, in this case TF is not using allgather, because MirroredStrategy uses legacy nccl ops. 
 MirroredStrategy could switch to use collective ops, which support allgather.  FYI @guptapriya @yuefengz 

If the tensor is small and can fit in memory, converting to a dense tensor is worth trying as Bairen suggested above.

**Comment by guptapriya** [comment]
@dubey yes we have considered doing this - for some cases, using the collective ops allgather would be more performant and in some cases, converting to dense tensor might actually be faster. We have not had the bandwidth to prioritize this yet, though.

**Comment by ratovarius** [comment]
Hi,
Similar issue here, 

I have this warning when running a keras multiple input model using functional API. The model works fine and without warnings when runing on single GPU. When I use `tf.distribute.MirroredStrategy` to use two GPUs the final result of the model is ok but I get the warning. I think this can lead to performance issues?

```
tf.__version__ : 2.2.0
tf.keras.__version__ : 2.3.0-tf
NVIDIA-SMI 410.72       Driver Version: 410.72       CUDA Version: 10.1

```
The model I am generating is:

```
def build_model_():

input_a_size = 200
input_b_size = 4
num_classes = 2
len_embedding = 100

mirrored_strategy = tf.distribute.MirroredStrategy(['/gpu:0', '/gpu:1'])

with mirrored_strategy.scope():

    input_a = Input(shape=(input_a_size,), name='input_a', dtype=np.uint8)
    input_b = Input(shape=(input_b_size,), name='input_b', dtype=np.float32)

    x = Embedding(len_embedding, 100)(input_a)
    x = Conv1D(128, 4, activation='relu')(x)
    x = MaxPooling1D(4)(x)
    x = Flatten()(x)
    branch_a = Dense(64, activation='relu')(x)

    x = Dense(32, activation='relu')(input_b)
    branch_b = Dense(32, activation='relu')(x)

    concat = Concatenate()([
                            branch_a,
                            branch_b,
                           ])

    x = Dense(256, activation = 'relu')(concat)
    output = Dense(num_classes, activation='softmax')(x)

    model = Model(inputs=[
                          input_a,
                          input_b,
                         ],
                  outputs=[output])
        
    model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

model.summary()

return model
```

Model summary:

```
INFO:tensorflow:Using MirroredStrategy with devices ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1')
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_a (InputLayer)            [(None, 200)]        0                                            
__________________________________________________________________________________________________
embedding (Embedding)           (None, 200, 100)     10000       input_a[0][0]                    
__________________________________________________________________________________________________
conv1d (Conv1D)                 (None, 197, 128)     51328       embedding[0][0]                  
__________________________________________________________________________________________________
max_pooling1d (MaxPooling1D)    (None, 49, 128)      0           conv1d[0][0]                     
__________________________________________________________________________________________________
input_b (InputLayer)            [(None, 4)]          0                                            
__________________________________________________________________________________________________
flatten (Flatten)               (None, 6272)         0           max_pooling1d[0][0]              
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 32)           160         input_b[0][0]                    
__________________________________________________________________________________________________
dense (Dense)                   (None, 64)           401472      flatten[0][0]                    
__________________________________________________________________________________________________
dense_2 (Dense)                 (None, 32)           1056        dense_1[0][0]                    
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 96)           0           dense[0][0]                      
                                                                 dense_2[0][0]                    
__________________________________________________________________________________________________
dense_3 (Dense)                 (None, 256)          24832       concatenate[0][0]                
__________________________________________________________________________________________________
dense_4 (Dense)                 (None, 2)            514         dense_3[0][0]                    
==================================================================================================
Total params: 489,362
Trainable params: 489,362
Non-trainable params: 0
__________________________________________________________________________________________________
```

The way I am generating the inputs:

```
input_a_train.shape: (35000, 200)
input_b_train.shape: (35000, 4)
y_train.shape: (35000, 2)
```

```
train_dataset = tf.data.Dataset.from_tensor_slices(({
                                                     "input_a": input_a_train,
                                                     "input_b": input_b_train,
                                                     }, y_train))
<TensorSliceDataset shapes: ({input_a: (200,), input_b: (4,)}, (2,)), types: ({input_a: tf.uint8, input_b: tf.float64}, tf.float32)>

val_dataset = tf.data.Dataset.from_tensor_slices(({
                                                     "input_a": input_a_val,
                                                     "input_b": input_b_val,
                                                     }, y_val))
<TensorSliceDataset shapes: ({input_a: (200,), input_b: (4,)}, (2,)), types: ({input_a: tf.uint8, input_b: tf.float64}, tf.float32)>

train_batches = train_dataset.padded_batch(128)
val_batches = val_dataset.padded_batch(128)
```

I get the warning during the training phase,

```
history = my_model.fit(
    x = train_batches,
    epochs=3,
    verbose = 1,
    validation_data = val_batches,
)
```

Here is the output:

```
Epoch 1/3
INFO:tensorflow:batch_all_reduce: 12 all-reduces with algorithm = nccl, num_packs = 1
WARNING:tensorflow:Efficient allreduce is not supported for 1 IndexedSlices
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:batch_all_reduce: 12 all-reduces with algorithm = nccl, num_packs = 1
WARNING:tensorflow:Efficient allreduce is not supported for 1 IndexedSlices
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:GPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:GPU:0', '/job:localhost/replica:0/task:0/device:GPU:1').
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
INFO:tensorflow:Reduce to /job:localhost/replica:0/task:0/device:CPU:0 then broadcast to ('/job:localhost/replica:0/task:0/device:CPU:0',).
274/274 [==============================] - ETA: 0s - loss: 0.1857 - accuracy: 0.9324
...
```

**Comment by nikitamaia** [comment]
Closing this thread since this is a known performance issue, and a workaround was found in #41898. If the workaround of using `MultiWorkerMirroredStrategy` explained in #41898 does not help your performance, please open a new issue.

**Comment by marchss** [comment]
> IndexedSlice are sparse data structures, so TF is using allgather for AllReduce. It probably doesn’t matter if you are only using 2 GPUs, but you could consider convert sparse tensor to a dense one for better AllReduce support.
> 
> cc @dubey

could you please clarify which tensor you are referring to should be converted from sparse to dense tensors in the example code? If convert sparse tensors to dense tensors, will this cause Out of Memory issue? Thanks.
